src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -133,12 +133,25 @@ private bool IsIrrefutablePatternTest(BoundExpression test)
 
         private BoundExpression CompareWithConstant(BoundExpression input, BoundExpression boundConstant)
         {
-            return _factory.StaticCall(
-                _factory.SpecialType(SpecialType.System_Object),
-                "Equals",
-                _factory.Convert(_factory.SpecialType(SpecialType.System_Object), boundConstant),
-                _factory.Convert(_factory.SpecialType(SpecialType.System_Object), input)
-                );
+            var systemObject = _factory.SpecialType(SpecialType.System_Object);
+            if (boundConstant is BoundLiteral bl && bl.ConstantValue.IsNull)
+            {
+                var systemBoolean = _factory.SpecialType(SpecialType.System_Boolean);
+                return _factory.Binary(
+                    kind: BinaryOperatorKind.Equal,
+                    type: systemBoolean,
+                    left: _factory.Convert(systemObject, input),
```

Comment:
I think you can have here:
BinaryOperatorKind.NullableNullEqual
BinaryOperatorKind.ObjectEquals
BinaryOperatorKind.OperatorMethodEquals     (with Object.Equals as method)